keanu-project/src/main/java/io/improbable/keanu/plating/loop/LoopBuilder.java

```diff
@@ -0,0 +1,180 @@
+import io.improbable.keanu.vertices.generic.nonprobabilistic.If;
+
+public class LoopBuilder {
+    private final Collection<Vertex> initialState;
+    private ImmutableMap.Builder<VertexLabel, VertexLabel> customMappings = ImmutableMap.builder();
+    private int maxLoopCount = Loop.DEFAULT_MAX_COUNT;
+    private boolean throwWhenMaxCountIsReached = true;
+
+    <V extends Vertex<?>> LoopBuilder(Collection<V> initialState) {
+        this.initialState = ImmutableList.copyOf(initialState);
+    }
+
+    /**
+     * An optional method to override the default value
+     *
+     * @param maxCount - the max number of times the loop can run
+     * @return self
+     */
+    public LoopBuilder atMost(int maxCount) {
+        this.maxLoopCount = maxCount;
+        return this;
+    }
+
+    /**
+     * An optional method to override the default behaviour
+     * If the maximum loop count is exceeded, it will log a warning instead of throwing
+     *
+     * @return self
+     */
+    public LoopBuilder dontThrowWhenMaxCountIsReached() {
+        this.throwWhenMaxCountIsReached = false;
+        return this;
+    }
+
+    /**
+     * An optional method to add custom mappings
+     *
+     * @param inputLabel  - the label assigned to the ProxyVertex in frame t
+     * @param outputLabel - the label assigned to a Vertex in frame t-1 which will become the ProxyVertex's parent
+     * @return self
+     */
+    public LoopBuilder mapping(VertexLabel inputLabel, VertexLabel outputLabel) {
+        customMappings.put(inputLabel, outputLabel);
+        return this;
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionSupplier - a lambda that creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Supplier<BoolVertex> conditionSupplier){
+        return whilst(plate -> conditionSupplier.get());
+    }
+
+    /**
+     * A mandatory method to specify the condition
+     *
+     * @param conditionFunction - a lambda that takes the current Plate and creates and returns a new BoolVertex
+     * @return the next stage builder
+     */
+    public LoopBuilder2 whilst(Function<Plate, BoolVertex> conditionFunction){
+        return new LoopBuilder2(initialState, conditionFunction, customMappings.build(), maxLoopCount, throwWhenMaxCountIsReached);
+    }
+
+    public class LoopBuilder2 {
+        private final Collection<Vertex> initialState;
+        private final Function<Plate, BoolVertex> conditionFunction;
+        private final Map<VertexLabel, VertexLabel> customMappings;
+        private final int maxLoopCount;
+        private final boolean throwWhenMaxCountIsReached;
+        private final VertexLabel VALUE_OUT_WHEN_ALWAYS_TRUE_LABEL = new VertexLabel("loop_value_out_when_always_true");
+        private final VertexLabel VALUE_IN_WHEN_ALWAYS_TRUE_LABEL = PlateBuilder.proxyFor(VALUE_OUT_WHEN_ALWAYS_TRUE_LABEL);
```

Comment:
Does the use of these labels cause problems if you have two loops in a single network?

Reply:
no, because everything is scoped by a unique plate ID, and all this is needed for is to join together vertices in adjacent plates.